### ISSUE TITLE
EFLAGS R/W Information

### ISSUE THREAD

**Issue by fvrmatteo** [capstone-engine/capstone#704]
I was looking into Capstone headers to know if there is a way to determine which flags an instruction reads/writes. I know there is "eflags" in the "x86" structure, but I can't find a way to determine which flags are read and which written. Am I missing something?

I think this can be a good feature because, at least in my case, it's fundamental to know everything possible about the EFLAGS register.

If needed I can take some time to check the correctness of EFLAGS information associated to each instruction.

**Comment by aquynh** [comment]
you should use the "next" branch, which can report the flags read/written by each instructions. if you see something wrong, please report.

**Comment by fvrmatteo** [comment]
Maybe I explained wrong. I was wondering if I can know "which flags are written" and "which flags are read"? I know 'next' has x86->flags, but that doesn't specify which flag was read, written or read+written. It is only a "list" of flags read/written but there is no way to determine specifically the access type of a flag. If I'm missing something let me know.

CMOVcc reads some flags and based on flags value executes a MOV instruction or not. But this instruction does not affect flags. I currently have no way to determine if x86->eflags contains flags which are written or read or both.

So, for example:

"CMOVZ r16, r/m16" = Move if zero (ZF=1). In the Capstone structure I'm expecting to see the flag ZF contained in "x86->eflags" with specified something like ZF_IS_READ about that specific bit-flag because ZF is only read, not updated or written.

"TEST r/m16,r16" = Logical Compare. From the Intel manual we know that: "The OF and CF flags are set to 0. The SF, ZF, and PF flags are set according to the result (see the "Operation" section above). The state of the AF flag is undefined.". In the Capstone structure I'm expecting to see the flags OF, CF, SF, ZF, PF contained in "x86->eflags" with specified something like OF_IS_WRITE, CF_IS_WRITE, SF_IS_WRITE, ZF_IS_WRITE, PF_IS_WRITE.

Let me know if I should explain it better :)

**Comment by aquynh** [comment]
we have bunch of related flags in https://github.com/aquynh/capstone/blob/next/include/capstone/x86.h#L73-L118, but if you want something unavailable there, feel free to send in a PR to add them.

thanks.

**Comment by fvrmatteo** [comment]
I didn't realize those flags were the one used by x86->eflags, sorry. This is exactly what I was looking for. If needed I will add something but the one I needed are already there :D

**Comment by fvrmatteo** [comment]
I'm reading the Intel manual now, and comparing EFLAGS information given by Capstone Next with information contained in the manual. I noticed something I can't fully understand by Capstone x86.h header.

I disassembled with Capstone Next the following example:

Input Instruction: bt ebp, eax
Output: [EFLAGS]: MOD_CF, UNDEF_OF, UNDEF_SF, UNDEF_ZF, UNDEF_PF, UNDEF_AF

Reading the manual for the instruction "BT = Bit Test" we have the following: "The CF flag contains the value of the selected bit. The ZF flag is unaffected. The OF, SF, AF, and PF flags are undefined."

CF, OF, SF, AF, PF are marked correctly, ZF is marked wrong (I think) because I was expecting to see "PRIOR_ZF" because "ZF is unaffected". If you think I'm right about it I will fix EFLAGS of various instructions :)
